### ISSUE TITLE
Test is failing with Karma but not with Wallaby

### ISSUE THREAD

**Issue by IT-CASADO** [wallabyjs/public#2607]
### Issue description or question

Wallaby does NOT show failing test.
But same test is failing with Karma (locally and in CI pipeline): 

To reproduce my issue:

1. Clone this [repository](https://github.com/IT-CASADO/wallaby-not-failing)
2. Run "yarn test" (Karma) -> you will see one failing test
3. Start Wallaby -> the same test does **NOT** fail

I'm expecting that the test is also failing with Wallaby!

### Wallaby diagnostics report

```
{
  editorVersion: '1.52.1',
  pluginVersion: '1.0.258',
  editorType: 'VSCode',
  osVersion: 'win32 10.0.19042',
  nodeVersion: 'v12.17.0',
  coreVersion: '1.0.1004',
  checksum: 'MjRmYjBmYzgwMjhkY2RmZDIwMTI0ZWIxYTlkM2EyMGEsMTYyNjA0ODAwMDAwMCww',
  config: {
    diagnostics: {
      angular: {
        workspace: {
          '$schema': './node_modules/@angular/cli/lib/config/schema.json',
          version: 1,
          newProjectRoot: 'projects',
          projects: {
            'wallaby-not-failing': {
              projectType: 'application',
              schematics: { '@schematics/angular:component': { style: 'scss' }, '@schematics/angular:application': { strict: true } },
              root: '',
              sourceRoot: 'src',
              prefix: 'app',
              architect: {
                build: {
                  builder: '@angular-devkit/build-angular:browser',
                  options: {
                    outputPath: 'dist/wallaby-not-failing',
                    index: 'src/index.html',
                    main: 'src/main.ts',
                    polyfills: 'src/polyfills.ts',
                    tsConfig: 'tsconfig.app.json',
                    aot: true,
                    assets: [ 'src/favicon.ico', 'src/assets' ],
                    styles: [ 'src/styles.scss' ],
                    scripts: []
                  },
                  configurations: {
                    production: {
                      fileReplacements: [ { replace: 'src/environments/environment.ts', with: 'src/environments/environment.prod.ts' } ],
                      optimization: true,
                      outputHashing: 'all',
                      sourceMap: false,
                      namedChunks: false,
                      extractLicenses: true,
                      vendorChunk: false,
                      buildOptimizer: true,
                      budgets: [ { type: 'initial', maximumWarning: '500kb', maximumError: '1mb' }, { type: 'anyComponentStyle', maximumWarning: '2kb', maximumError: '4kb' } ]
                    }
                  }
                },
                serve: {
                  builder: '@angular-devkit/build-angular:dev-server',
                  options: { browserTarget: 'wallaby-not-failing:build' },
                  configurations: { production: { browserTarget: 'wallaby-not-failing:build:production' } }
                },
                'extract-i18n': { builder: '@angular-devkit/build-angular:extract-i18n', options: { browserTarget: 'wallaby-not-failing:build' } },
                test: {
                  builder: '@angular-devkit/build-angular:karma',
                  options: {
                    main: 'src/test.ts',
                    polyfills: 'src/polyfills.ts',
                    tsConfig: 'tsconfig.spec.json',
                    karmaConfig: 'karma.conf.js',
                    assets: [ 'src/favicon.ico', 'src/assets' ],
                    styles: [ 'src/styles.scss' ],
                    scripts: []
                  }
                },
                lint: {
                  builder: '@angular-devkit/build-angular:tslint',
                  options: { tsConfig: [ 'tsconfig.app.json', 'tsconfig.spec.json', 'e2e/tsconfig.json' ], exclude: [ '**/node_modules/**' ] }
                },
                e2e: {
                  builder: '@angular-devkit/build-angular:protractor',
                  options: { protractorConfig: 'e2e/protractor.conf.js', devServerTarget: 'wallaby-not-failing:serve' },
                  configurations: { production: { devServerTarget: 'wallaby-not-failing:serve:production' } }
                }
              }
            }
          },
          defaultProject: 'wallaby-not-failing'
        },
        main: '// This file is required by karma.conf.js and loads recursively all the .spec and framework files\n' +
          '\n' +
          "import 'zone.js/dist/zone-testing';\n" +
          "import { getTestBed } from '@angular/core/testing';\n" +
          'import {\n' +
          '  BrowserDynamicTestingModule,\n' +
          '  platformBrowserDynamicTesting\n' +
          "} from '@angular/platform-browser-dynamic/testing';\n" +
          '\n' +
          'declare const require: {\n' +
          '  context(path: string, deep?: boolean, filter?: RegExp): {\n' +
          '    keys(): string[];\n' +
          '    <T>(id: string): T;\n' +
          '  };\n' +
          '};\n' +
          '\n' +
          '// First, initialize the Angular testing environment.\n' +
          'getTestBed().initTestEnvironment(\n' +
          '  BrowserDynamicTestingModule,\n' +
          '  platformBrowserDynamicTesting()\n' +
          ');\n' +
          '// Then we find all the tests.\n' +
          '\n' +
          '// And load the modules.\n' +
          '\n'
      }
    },
    testFramework: { version: 'jasmine@3.5.0', configurator: 'jasmine@2.1.3', reporter: 'jasmine@2.1.3', starter: 'jasmine@2.1.3', autoDetected: true },
    env: { kind: 'chrome', type: 'browser', params: {}, viewportSize: { width: 800, height: 600 }, options: { width: 800, height: 600 }, bundle: true },
    files: [
      { pattern: 'src/**/*.spec.ts', ignore: true, trigger: true, load: true },
      { pattern: 'src/polyfills.ts', ignore: true, trigger: true, load: true },
      { pattern: 'src/test.ts', ignore: true, trigger: true, load: true },
      { pattern: 'src/main.ts', ignore: true, trigger: true, load: true },
      { pattern: 'src/index.html', ignore: true, trigger: true, load: true },
      { pattern: 'src/test.base.ts', ignore: true, trigger: true, load: true },
      { pattern: 'src/test.wallaby.ts', ignore: false, trigger: true, load: true, order: 1 },
      { pattern: 'src/**/*.+(ts|js)', load: false, ignore: false, trigger: true, order: 2 },
      { pattern: 'src/**/*.+(css|less|scss|sass|styl|html|json|svg)', instrument: false, load: false, ignore: false, trigger: true, order: 3 }
    ],
    tests: [ { pattern: 'src/**/*.spec.ts', load: false, ignore: false, trigger: true, test: true, order: 4 } ],
    filesWithNoCoverageCalculated: [],
    runAllTestsInAffectedTestFile: false,
    updateNoMoreThanOneSnapshotPerTestFileRun: false,
    compilers: {},
    preprocessors: {},
    maxConsoleMessagesPerTest: 100,
    autoConsoleLog: true,
    delays: { run: 0, edit: 100, update: 0 },
    workers: { initial: 0, regular: 0, recycle: false },
    teardown: undefined,
    hints: {
      ignoreCoverage: '__REGEXP /ignore coverage|istanbul ignore/',
      ignoreCoverageForFile: '__REGEXP /ignore file coverage/',
      commentAutoLog: '?',
      testFileSelection: { include: '__REGEXP /file\\.only/', exclude: '__REGEXP /file\\.skip/' }
    },
    automaticTestFileSelection: true,
    runSelectedTestsOnly: false,
    extensions: {},
    reportUnhandledPromises: false,
    throwOnBeforeUnload: true,
    slowTestThreshold: 75,
    lowCoverageThreshold: 80,
    loose: true,
    configCode: 'auto.detect#1313788472'
  },
  packageJSON: {
    dependencies: {
      '@angular/animations': '~11.0.0',
      '@angular/common': '~11.0.0',
      '@angular/compiler': '~11.0.0',
      '@angular/core': '~11.0.0',
      '@angular/forms': '~11.0.0',
      '@angular/platform-browser': '~11.0.0',
      '@angular/platform-browser-dynamic': '~11.0.0',
      '@angular/router': '~11.0.0',
      '@ngrx/store': '10.1.2',
      '@ngrx/effects': '10.1.2',
      '@ngrx/router-store': '10.1.2',
      '@ngrx/component-store': '10.1.2',
      rxjs: '~6.6.0',
      tslib: '^2.0.0',
      'zone.js': '~0.10.2'
    },
    devDependencies: {
      '@angular-devkit/build-angular': '~0.1100.1',
      '@angular/cli': '~11.0.1',
      '@angular/compiler-cli': '~11.0.0',
      '@types/jasmine': '~3.6.0',
      '@types/node': '^12.11.1',
      codelyzer: '^6.0.0',
      'jasmine-core': '~3.6.0',
      'jasmine-spec-reporter': '~5.0.0',
      karma: '~5.1.0',
      'karma-chrome-launcher': '~3.1.0',
      'karma-coverage': '~2.0.3',
      'karma-jasmine': '~4.0.0',
      'karma-jasmine-html-reporter': '^1.5.0',
      protractor: '~7.0.0',
      'ts-node': '~8.3.0',
      tslint: '~6.1.0',
      typescript: '~4.0.2'
    }
  },
  fs: { numberOfFiles: 15 },
  debug: [
    '2021-01-17T19:50:59.828Z angular/cli config Detected Angular CLI.\n',
    '2021-01-17T19:51:03.321Z angular/cli karma project config Using virtual T:/GITHUBMY/wallaby-not-failing/src/test.ts.\n',
    '2021-01-17T19:51:03.369Z project Wallaby Node version: v12.17.0\n',
    '2021-01-17T19:51:03.369Z project Wallaby config: T:\\GITHUBMY\\wallaby-not-failing\\auto.detect\n',
    '2021-01-17T19:51:04.142Z project File cache: C:\\Users\\danie\\.vscode\\extensions\\wallabyjs.wallaby-vscode-1.0.258\\projects\\ebcd40787c073a86\n',
    '2021-01-17T19:51:04.226Z uiService Listening port 51235\n',
    '2021-01-17T19:51:04.301Z workers Parallelism for initial run: 2, for regular run: 2\n',
    '2021-01-17T19:51:04.301Z workers Starting run worker instance #0\n',
    '2021-01-17T19:51:04.301Z workers Starting run worker instance #1\n',
    '2021-01-17T19:51:04.305Z workers Web server is listening at 62315\n',
    '2021-01-17T19:51:04.314Z project Stopping process pool\n',
    '2021-01-17T19:51:04.314Z project File cache is up-to-date, starting full test run\n',
    '2021-01-17T19:51:04.316Z project Running postprocessor\n',
    '2021-01-17T19:51:04.319Z postprocessor Processing started.\n',
    '2021-01-17T19:51:04.319Z postprocessor Initializing.\n',
    '2021-01-17T19:51:04.319Z postprocessor Initializing entry.\n',
    '2021-01-17T19:51:04.320Z postprocessor Sanitizing configuration.\n',
    '2021-01-17T19:51:04.320Z postprocessor Adjusting configuration.\n',
    '2021-01-17T19:51:04.320Z postprocessor Creating compiler.\n',
    '2021-01-17T19:51:04.549Z postprocessor Customizing compiler.\n',
    '2021-01-17T19:51:04.549Z postprocessor Compilation started.\n',
    '2021-01-17T19:51:10.074Z workers Started run worker instance (immediate) #0\n',
    '2021-01-17T19:51:10.074Z workers Started run worker instance (immediate) #1\n',
    '2021-01-17T19:51:14.639Z postprocessor Compilation finished.\n',
    '2021-01-17T19:51:15.429Z postprocessor Emitting 334 files.\n',
    '2021-01-17T19:51:15.433Z postprocessor Processing finished.\n',
    '2021-01-17T19:51:15.474Z project Postprocessor execution finished\n',
    '2021-01-17T19:51:15.475Z project Test run started; run priority: 3\n',
    '2021-01-17T19:51:15.476Z project Running all tests\n',
    '2021-01-17T19:51:15.488Z workers Starting test run, priority: 3\n',
    '2021-01-17T19:51:15.488Z workers Distributing tests between 2 workers\n',
    '2021-01-17T19:51:15.489Z workers Running tests in parallel\n',
    '2021-01-17T19:51:15.489Z chromeRunner Starting sandbox [worker #0, session #ayccs]\n',
    '2021-01-17T19:51:15.489Z chromeRunner Starting sandbox [worker #1, session #t3flm]\n',
    '2021-01-17T19:51:15.490Z chromeRunner Preparing sandbox [worker #0, session #ayccs]\n',
    '2021-01-17T19:51:15.493Z browserRunner Total files to load in sandbox: 5\n',
    '2021-01-17T19:51:15.494Z browserRunner Sandbox is generated [worker #0, session #ayccs]: http://localhost:62315/wallaby_sandbox0.html\n',
    '2021-01-17T19:51:15.494Z chromeRunner Preparing sandbox [worker #1, session #t3flm]\n',
    '2021-01-17T19:51:15.497Z browserRunner Total files to load in sandbox: 5\n',
    '2021-01-17T19:51:15.498Z browserRunner Sandbox is generated [worker #1, session #t3flm]: http://localhost:62315/wallaby_sandbox1.html\n',
    '2021-01-17T19:51:15.498Z chromeRunner Prepared sandbox [worker #0, session #ayccs]\n',
    '2021-01-17T19:51:15.498Z chromeRunner Prepared sandbox [worker #1, session #t3flm]\n',
    '2021-01-17T19:51:15.498Z workers [worker #0, session #ayccs] Running tests in sandbox\n',
    '2021-01-17T19:51:15.500Z workers [worker #1, session #t3flm] Running tests in sandbox\n',
    '2021-01-17T19:51:15.997Z workers [ayccs] Loaded 0 test(s)\n',
    '2021-01-17T19:51:16.003Z workers [ayccs] Run 0 test(s), skipped 0 test(s)\n',
    '2021-01-17T19:51:16.005Z workers [ayccs] Sandbox is responsive, closing it\n',
    '2021-01-17T19:51:16.039Z workers [t3flm] Loaded 1 test(s)\n',
    '2021-01-17T19:51:16.108Z workers [t3flm] Test executed: should create the app\n',
    '2021-01-17T19:51:16.159Z workers [t3flm] Run 1 test(s), skipped 0 test(s)\n',
    '2021-01-17T19:51:16.160Z workers [t3flm] Sandbox is responsive, closing it\n',
    '2021-01-17T19:51:16.162Z workers Merging parallel test run results\n',
    '2021-01-17T19:51:16.163Z project Test run finished\n',
    '2021-01-17T19:51:16.163Z project Processed console.log entries\n',
    '2021-01-17T19:51:16.164Z project Processed loading sequences\n',
    '2021-01-17T19:51:16.164Z project Processed executed tests\n',
    '2021-01-17T19:51:16.165Z project Processed code coverage\n',
    '2021-01-17T19:51:16.174Z project Test run result processed and sent to IDE\n'
  ]
}
```

**Comment by smcenlly** [comment]
Thanks for reporting the error. Your test is failing in Angular's default `afterAll` and for some reason Wallaby isn't detecting this. At this point, I'm not yet sure if this is intentional (by design) or a bug.

At any rate, the problem is that your code is not setting the initialState of your mock store in your `provideMockStore` call. You will need to initialize it with a valid state for your exported selectors in `src/app/counter.reducer.ts`. You can do this by changing your `beforeEach` function in `src/app/my-counter/my-counter.component.spec.ts` to:

```diff
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [MyCounterComponent],
      providers: [
-       provideMockStore()
+       provideMockStore({
+         initialState: {
+           count: {
+             counter: 0
+           }
+         }
+       })
      ],
    }).compileComponents();
  });
```

We will post back with an update once we work out whether this behavior is intentional or not.

**Comment by IT-CASADO** [comment]
For me this is a bug in Wallaby.

As you already mentioned, I could get the test to work by providing an initial state to the mock store. 

But by not providing an initial state (intentionally or not), I'm expecting to see an exception on accessing a property on an uninitialised state. Karma is reporting this exception to me... (and therefore I immediately get noticed that I forgot something... )

**Comment by smcenlly** [comment]
> For me this is a bug in Wallaby.

What you're seeing is expected behavior and is not a bug in Wallaby. The problem that you're having is caused by a race condition from your asynchronous code and you can and will get the same error when using Karma. I've included a short demonstration below on your project to show you (see animated gif). In this case, I simulated having a second long running test in the same spec file to make it easier to duplicate the problem.

Because of execution timing, Wallaby seems to not report the error more often than not (in your example) while Karma does report the error more often than not. This is probably in part because of how aggressive Wallaby is in returning results as soon as they're ready but is also simply because of differences in execution time and differences in the internals of Wallaby and Karma. Because this is a race condition, depending on your test running and test structure, you may see this change to the other way around. Interestingly, the problem exhibits in a number of different ways depending on exactly when the error occurs in the context of everything else running. Sometimes an error will be reported, other times not. It's also possible for the error to be incorrectly reported against a different (seemingly random test).

In your case, the problem is that your observable subscriptions are asynchronous and the bug that is causing an exception to occur happens **after** the test run has finished. If you want your Wallaby (and Karma) to handle your race condition then you will need to change your tests to be aware of your subscriptions. Perhaps something like [this](https://medium.com/angular-in-depth/how-to-test-observables-a00038c7faad) or [this](https://itnext.io/a-quick-tip-on-testing-observables-e2fbdebef4c) medium post?

![yarn_test_not_reporting_error](https://user-images.githubusercontent.com/2074089/104880537-74df4100-59ab-11eb-81fd-4b1322c6b439.gif)

**Comment by IT-CASADO** [comment]
@smcenlly Thanks for your detail response to my issue!

I didn't expect an **async** observable here. So my assumption was that everything is running synchronously...

**Comment by IT-CASADO** [comment]
Duplicate of #2063
